src/slash_commands/boardowa.rs

```diff
@@ -142,105 +105,99 @@ impl SlashCommand for BoardowaCommand {
+            opening.from,
+            opening.to
+        );
+    }
+    Ok(opening)
 }
 
-impl TableInfo {
-    fn availability_ranges(&self) -> Vec<AvailabilityRange> {
-        let mut ret = vec![];
-        let mut current_range = None;
-        for (idx, av) in self.availability.iter().enumerate() {
-            if !av {
-                if let Some(r) = current_range {
-                    ret.push(r);
-                    current_range = None;
-                }
-            } else {
-                match &mut current_range {
-                    None => {
-                        current_range = Some(AvailabilityRange {
-                            table_id: self.value.clone(),
-                            start: idx,
-                            len: 1,
-                        });
-                    }
-                    Some(r) => {
-                        r.len += 1;
-                    }
-                }
-            }
-        }
-        if let Some(r) = current_range {
-            ret.push(r);
-        }
-        ret
-    }
+fn build_message(
+    rank: Vec<AvailabilityRange>,
+    date: NaiveDate,
+    opening: TimeRange,
+    n_tables: usize,
+) -> String {
+    let mut message = format!(
+        "Consider following tables for {}:\n```\n",
+        date.format("%Y-%m-%d")
+    );
+    message.push_str(
+        &rank
+            .into_iter()
+            .take(n_tables)
+            .map(|range| format_table(&range, &opening))
+            .collect::<Vec<String>>()
+            .join("\n"),
+    );
+    message.push_str("\n```");
+    message
 }
 
-fn transpose(availability: Vec<Vec<HourlyTableInfo>>) -> Result<Vec<TableInfo>, Box<dyn Error>> {
-    let mut map = HashMap::new();
-    if let Some(tables) = availability.first() {
-        for table in tables {
-            map.insert(
-                &table.value,
-                TableInfo {
-                    value: table.value.clone(),
-                    availability: vec![],
-                },
-            );
-        }
-    }
-    for tables in &availability {
-        if tables.len() != map.len() {
-            return Err(Box::new(BoardowaError(String::from(
-                "Internal error: unexpected table",
-            ))));
-        }
-        for table in tables {
-            map.get_mut(&table.value)
-                .ok_or("Missing table")?
-                .availability
-                .push(table.available);
-        }
-    }
-    Ok(map.into_values().collect())
+fn format_table(range: &AvailabilityRange, opening: &TimeRange) -> String {
+    format!(
+        "Table {} | {} ppl. | {}-{}",
+        replace_leading_zeros(range.table_id.as_str()),
+        range.capacity,
+        format_time(opening.from.into(), range.start),
+        format_time(opening.from.into(), range.start + (range.len as usize))
+    )
 }
 
-fn rank_tables(tables: Vec<TableInfo>) -> Vec<AvailabilityRange> {
```

Comment:
What is the purpose of this function? Just remove leading zeros and replace them with spaces? Maybe just [trim_start_matches](https://doc.rust-lang.org/std/primitive.str.html#method.trim_start_matches) the original string and add n times space, where n is the difference between new and old string lengths?

Reply:
Woah, didn't notice this function. Love this language

Reply:
(the purpose of this function was użycie `Cow` bo bardzo mi się podoba)